File: Python/18.4Sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        n, res = len(nums), []
        nums.sort()
        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            if i + 3 < n - 1:
                if nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target:
                    break
            if i < n - 3:
                if nums[i] + nums[n - 1] + nums[n - 2] + nums[n - 3] < target:
                    continue
            for j in range(i + 1, n):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                l, r = j + 1, n - 1
                while l < r:
                    tmp = nums[i] + nums[j] + nums[l] + nums[r]
                    if tmp == target:
                        res.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l + 1]:
                            l += 1
                        while l < r and nums[r] == nums[r - 1]:
                            r -= 1
                        l += 1
                        r -= 1
                    elif tmp > target:
                        r -= 1
                    else:
                        l += 1
        return res

File: Javascript/18.4Sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    let n = nums.length;
    if (nums == null || n < 4) {
        return [];
    }
    let num = nums.sort((a, b) => a - b);
    let res = []
    for (let i = 0; i < n - 3; i++) {
        if (i > 0 && num[i] === num[i - 1]) {
            continue;
        }
        if (num[i] + num[i + 1] + num[i + 2] + num[i + 3] > target) {
            break;
        }
        if (num[i] + num[n - 1] + num[n - 2] + num[n - 3] < target) {
            continue;
        }
        for (let j = i + 1; j < n - 2; j++) {
            let left = j + 1;
            let right = n - 1;
            if (j > i + 1 && num[j] === num[j - 1]) {
                continue;
            }
            while (left < right) {
                if (num[i] + num[j] + num[left] + num[right] === target) {
                    res.push([num[i], num[j], num[left], num[right]]);
                    while (left < right && num[left] === num[left + 1]) {
                        left++;
                    }
                    while (left < right && num[right] === num[right - 1]) {
                        right--;
                    }
                    left++;
                    right--;
                }
                else {
                    num[i] + num[j] + num[left] + num[right] > target ? right-- : left++
                }
            }
        }
    }
    return res;
};